torch/fx/experimental/proxy_tensor.py

```diff
@@ -410,3 +410,47 @@ def wrap_fake(x):
         return t
 
     return wrapped
+
+
+def _find_proxy_tensor(*objects_to_search):
+    # return first proxy tensor found or None
+    proxy_tensors = (o for o in objects_to_search if isinstance(o, ProxyTensor))
+    return next(proxy_tensors, None)
+
+
+def get_isolated_graphmodule(func, args, kwargs):
+    """A helper function used to get the GraphModule for the given func.
+
+    It's expected to be used in the ProxyTensor tracing context.
+    It detaches the args and kwargs from the current tracer so that the trace of
+    the current graph module can be created without any side-effects.
+    """
+    # make_fx doesn't support kwargs, so we need to do this flattening
+    # and then unflatten the args before calling func
+    all_args, spec = pytree.tree_flatten((args, kwargs))
+
+    def wrapped(args):
+        fn_args, fn_kwargs = pytree.tree_unflatten(args, spec)
+        return func(*fn_args, **fn_kwargs)
+
+    # extract old tracer object
+    proxy_tensor = _find_proxy_tensor(*all_args)
+    if proxy_tensor is not None:
+        old_tracer = proxy_tensor.proxy.tracer
+
+    # create a new tracer object
+    graph = torch.fx.Graph()
+    new_tracer = PythonKeyTracer()
+    new_tracer.graph = graph
+
+    try:
+        for arg in all_args:
+            if isinstance(arg, ProxyTensor):
+                arg.proxy.tracer = new_tracer
+
+        gm = make_fx(wrapped)(all_args)
```

Comment:
This doesn't feel like it is enough. If there is an ambient proxy tensor mode active, it was never disabled and will still be interposing on invocations. But your test is passing. Do you have an explanation for why this is working as is?

Reply:
It looks like the answer is, you're in a context where the proxy tensor mode is disabled by the time you called this

Reply:
It works because all the calls are recorded on the different tracer which is created in this function.

Reply:
I see... this feels so bad haha

Reply:
Do you have a different understanding of the expected behavior? We have ProxyTensors as inputs to the function they all are expected to have the same tracer object attached (there's an assert for this somewhere), this tracer object is the place where the information about the calls gets stored and when we swap the tracer information about the calls is getting stored elsewhere.

Reply:
One reason it feels bad is because the ProxyTensorMode itself *contains a tracer*, and that tracer is now inconsistent with the tracers on the proxy objects. In fact, this means that if you have factory functions inside the conditional rewrite, they will go to the wrong graph. So this is definitely wrong!

Reply:
Yes, this is definitely wrong.
I added a test case with factory functions and two nested `get_isolated_graphmodule` calls. The test passes with:
```
mode = torch._C._get_torch_dispatch_mode()
with enable_torch_dispatch_mode(mode.inner, replace=mode):
    make_fx(...)
```
Then I added a test case with two nested `make_fx` and one `get_isolated_graphmodule` with factory functions. It also now passes with ExistStack of `enable_torch_dispatch_mode(mode.inner, replace=mode)` contexts.